# Evaluation terminale : Machine learning & python -- KSNATKIN Philippe

Cet examen se compose de 3 parties : 
<ol>
  <li>Restitution théorique</li>
  <li>Restitution pratique</li>
  <li>Cas pratique</li>
</ol>

## Partie 1 : Restitution théorique

Pour cette partie, l'objectif est de décrire des notions théoriques. Essayez de rédiger des réponses claires en utilisant vos mots.

#### Question 1.1: Donner une définition du machine learning. Eviter de recopier de wikipédia, essayer d'en donner une description en quelques mots.

Réponse : Le machine learning est une des méthodes de l’intelligence artificielle. Le machine learning permet à un système informatique d’apprendre et de s’adapter sans pour autant avoir reçu des instructions explicites. Cet apprentissage s’effectue en utilisant des algorithmes, des modèles statistiques et en analysant les paternes de la data. 

#### Question 1.2: Donner quelques applications concrètes d'utilisation du machine learning.

Réponse : Les utilisations sont multiples : les réseaux sociaux, la médecine, l’audio, les objets connectés, jeux vidéo, algorithmes de recommandations, la robotique, jusqu’à l’analyse de base de données (BIG DATA).   

#### Question 1.3: Quelles sont les 2 principales familles "d'apprentissage"? Donner une rapide définition de chacune

Réponse : Les deux types d’apprentissage sont le supervisé et le non supervisé. L’apprentissage supervisé consiste à définir soi même les entrées et les sorties, tandis que le non supervisé consiste à laisser l’algorithme créer lui-même les entrées et leur structure. 

#### Question 1.4: Dans laquelle de ces familles appartient la "classification"? Le "clustering"?

Réponse : Clustering : non supervisé. Classification : supervisé

#### Question 1.5: Définissez ce qu'est un arbre de décision et décrivez  de manière simple sa construction.

Réponse : Un arbre de décision est un schéma représentant les résultats possibles d'une série de choix interconnectés. Il peut être utilisé pour générer un algorithme qui détermine le meilleur choix de façon mathématique. Chaque nœud interne est un test sur une des dimension x, chaque branche est un résultat de test et chaque feuille est un label y. 

#### Question 1.6: Dans quelle situation parle-t-on "d'overfitting"?

Réponse : On parle d’overfitting quand le résultat correspond trop à une partie d’un ensemble de données en particulier. Le modèle s’adapte trop aux données d’entrainement ce qui augmente au global les marges d’erreurs. 

## Partie 2 : Etude d'un cas pratique

Pour cette partie, vous devrez vous inspirer des TP réalisés en cours. 

Il vous faut importer les librairies qui vous seront nécessaires répondre à la problèmatique dans la cellule suivante :

In [12]:
# packages
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier, plot_tree
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.neural_network import MLPClassifier
from sklearn.preprocessing import StandardScaler
from sklearn import metrics
import matplotlib.pyplot as plt

Importez les données comme suit :

In [3]:
data = pd.read_csv('nba_logreg.csv')

Ce jeu de données décrit les caracteristiques de jeu de joueurs de Basketball au sein de la NBA. La variable d'intérêt Y est "TARGET_5yrs". La description de ce dataset se trouve dans le fichier "NBA_variables_doc.png"

__L'objectif du problème est de réussir à prédire si un jeune joueur durera plus ou moins que 5ans au sein de la ligue.__

In [15]:
data

Name  GP   MIN  PTS  FGM  FGA   FG%  3P Made  3PA   3P%  ...  \
0      Brandon Ingram  36  27.4  7.4  2.6  7.6  34.7      0.5  2.1  25.0  ...   
1     Andrew Harrison  35  26.9  7.2  2.0  6.7  29.6      0.7  2.8  23.5  ...   
2      JaKarr Sampson  74  15.3  5.2  2.0  4.7  42.2      0.4  1.7  24.4  ...   
3         Malik Sealy  58  11.6  5.7  2.3  5.5  42.6      0.1  0.5  22.6  ...   
4         Matt Geiger  48  11.5  4.5  1.6  3.0  52.4      0.0  0.1   0.0  ...   
...               ...  ..   ...  ...  ...  ...   ...      ...  ...   ...  ...   
1335      Chris Smith  80  15.8  4.3  1.6  3.6  43.3      0.0  0.2  14.3  ...   
1336      Brent Price  68  12.6  3.9  1.5  4.1  35.8      0.1  0.7  16.7  ...   
1337     Marlon Maxey  43  12.1  5.4  2.2  3.9  55.0      0.0  0.0   0.0  ...   
1338  Litterial Green  52  12.0  4.5  1.7  3.8  43.9      0.0  0.2  10.0  ...   
1339        Jon Barry  47  11.7  4.4  1.6  4.4  36.9      0.4  1.3  33.3  ...   

      FTA   FT%  OREB  DREB  REB  AST  STL  BLK  TOV  TARGET_5Yrs  
0     2.3  69.9   0.7   3.4  4.1  1.9  0.4  0.4  1.3          0.0  
1     3.4  76.5   0.5   2.0  2.4  3.7  1.1  0.5  1.6          0.0  
2     1.3  67.0   0.5   1.7  2.2  1.0  0.5  0.3  1.0          0.0  
3     1.3  68.9   1.0   0.9  1.9  0.8  0.6  0.1  1.0          1.0  
4     1.9  67.4   1.0   1.5  2.5  0.3  0.3  0.4  0.8          1.0  
...   ...   ...   ...   ...  ...  ...  ...  ...  ...          ...  
1335  1.5  79.2   0.4   0.8  1.2  2.5  0.6  0.2  0.8          0.0  
1336  1.0  79.4   0.4   1.1  1.5  2.3  0.8  0.0  1.3          1.0  
1337  1.6  64.3   1.5   2.3  3.8  0.3  0.3  0.4  0.9          0.0  
1338  1.8  62.5   0.2   0.4  0.7  2.2  0.4  0.1  0.8          1.0  
1339  1.0  67.3   0.2   0.7  0.9  1.4  0.7  0.1  0.9          1.0  

[1340 rows x 21 columns]

#### Question 2.1: Donnez au moins une colonne qui doit être enlevé du jeu de données. Pourquoi? 

Réponse : On peut enlever la colonne "Name" car elle est inutile au problème posé. 

#### Exercice 2.1: Enlevez les variables du dataset qui vous semble inutile au problème.

In [1]:
# drop name
data.drop(columns=['Name'],axis=1, inplace=True)

NameError: name 'data' is not defined

#### Question 2.2: Ce dataset contient - il des données manquantes ou non exploitable?

data.isnull().sum()

Réponse : oui avant le drop de la colonne 3P%

#### Exercice 2.2: Si oui, donnez en la proportion par colonne.

In [17]:
((data.sum() - data.isnull().sum()) / data.sum()) * 100

#### Exercice 2.3: Séparez les données en 2 échantillons train-test

In [17]:
# code..
y = data['TARGET_5Yrs']
X = data.drop(labels=['TARGET_5Yrs'], axis=1)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state=42)

#### Question 2.3: Pourquoi cette étape est-elle indispensable?

Réponse : cette étape permet d'améliorer la précision de notre modèle

#### Exercice 2.4: Entrainez sur ces données un arbre de décision. Calculer une prédiction.

In [17]:
# code..
clf = DecisionTreeClassifier()
clf = clf.fit(X_train, Y_train)
# estimation
estimate = clf.predict_proba(X_test)[:,1]
true = Y_test
fpr, tpr, _ = roc_curve(Y_test,  estimate)
auc_ = auc(fpr, tpr)
print(auc_)

#### Exercice 2.5: Evaluez la qualité du modèle. 

In [18]:
# code..
accuracy_score(y_test, estimate, normalize=True, sample_weight=None)

#### Question 2.4: Quelle métrique utilisez vous? Pourquoi? 

Réponse : accuracy_score, c'est le plus simple etant donné les informations y_test et estimate dont on dispose facilement

#### Question 2.5: Le modèle construit est-il de qualité satisfaisante? Expliquez.

Réponse : Etant donné le résultat on peut douter de la qualité du modèle car le score est inférieur à 0

#### Exercice 2.6: Produisez une visualisation graphique de l'arbre construit

In [18]:
# code..
plt.figure(figsize=(17,11))
plot_tree(clf, feature_names = list(X_train.columns), fontsize=10)
plt.show()

#### Question 2.6: Donnez un exemple de lecture d'un tel arbre.

Réponse : Si GP <= 64.5 alors si OREB <= 0.85 ...

#### Question 2.7: Comment peut - on améliorer la qualité du modèle? Donnez au moins 3 méthodes:

Réponse : Les options pourraient être d'optimiser les hyperparamètres et d'utiliser des données et caractéristiques plus significatives. Enfin, il serait possible de faire une validation croisée. 

#### Exercice 2.7: Essayer d'appliquer une de ces méthodes

In [19]:
# code..
param_space = {'criterion':[ 'entropy', 'gini'], 
               'max_depth': list(range(2,10,1)),
               'min_samples_split':[0.01, 0.02, 0.03, 0.04, 0.05, 0.06, 0.07],
               'random_state': [42]}

clf_grid = DecisionTreeClassifier()
grid_search_tree = GridSearchCV(estimator=clf_grid,
                                param_grid=param_space,
                                scoring='roc_auc',
                                cv=8)
grid_search_tree.fit(X, Y.values)
best_parameters = grid_search_tree.best_params_
best_auc = grid_search_tree.best_score_

clf = DecisionTreeClassifier(criterion= 'gini', max_depth=3, min_samples_split=0.1, random_state=42)
clf = clf.fit(X_train, Y_train)
estimate = clf.predict_proba(X_test)[:,1]
true = Y_test
fpr, tpr, _ = roc_curve(Y_test,  estimate)
auc_ = auc(fpr, tpr)

# nouvel arbre
plt.figure(figsize=(17,11))
plot_tree(clf, feature_names = list(X_train.columns), fontsize=10)
plt.show()

#### Question 2.8: Concluez

Réponse : nous pouvons en conclure que seulement 1/2 personnes fera plus de 5 ans dans la ligue.